--- a/paper-server/build.gradle.kts
+++ b/paper-server/build.gradle.kts
@@ -17,11 +_,72 @@
     mache("io.papermc:mache:26.2+build.1")
     paperclip("io.papermc:paperclip:3.0.4")
 }
+// tessera 26.2 start - change output file
+val tesseraBuildVersion = providers
+    .gradleProperty("tesseraBuildVersion")
+    .orElse("LOCAL")
+    .get()
+
+val tesseraBuildChannel = providers
+    .gradleProperty("tesseraBuildChannel")
+    .orElse("DEV")
+    .get()
+
+val tesseraMinecraftVersion = providers
+    .gradleProperty("mcVersion")
+    .orElse("26.2")
+    .get()
+
+val invalidFileNameCharacters = Regex("[^A-Za-z0-9._-]")
+
+val safeBuildVersion = tesseraBuildVersion.replace(invalidFileNameCharacters, "-")
+val safeBuildChannel = tesseraBuildChannel.replace(invalidFileNameCharacters, "-")
+val safeMinecraftVersion = tesseraMinecraftVersion.replace(invalidFileNameCharacters, "-")
+val tesseraPaperclipFileName = "tessera-server-$safeMinecraftVersion.build-$safeBuildVersion-$safeBuildChannel.jar"
+
+tasks.createPaperclipJar {
+    outputZip.set(layout.buildDirectory.file("libs/$tesseraPaperclipFileName"))
+}
+//// settings for this are inside gradle.properties
+//val tesseraBuildVersion = providers.gradleProperty("tesseraBuildVersion").orElse("LOCAL")
+//val tesseraBuildChannel = providers.gradleProperty("tesseraBuildChannel").orElse("DEV")
+//val tmcVersion = providers.gradleProperty("mcVersion").orElse("26.2").get()
+//
+//val tesseraPaperclipPath = providers.zip(tesseraBuildVersion, tesseraBuildChannel)
+//{ buildNumber, buildChannel ->
+//    val safeBuildVersion = buildNumber.replace(Regex("[^A-Za-z0-9._-]"), "-")
+//    val safeBuildChannel = buildChannel.replace(Regex("[^A-Za-z0-9._-]"), "-")
+//
+//    val safeMCVersion = tmcVersion.replace(Regex("[^A-Za-z0-9._-]"), "-")
+//
+//    //tessera-paperclip-26.2.local-SNAPSHOT_old
+//    "libs/tessera-server-$safeMCVersion.build-$safeBuildVersion-$safeBuildChannel.jar"
+//}
+//
+//
+//tasks.createPaperclipJar {
+//    // set output to tessera Paper clip Path
+//    outputZip.set(layout.buildDirectory.file(tesseraPaperclipPath))
+//}
+// tessera 26.2 end - change output file
 
 paperweight {
     minecraftVersion = providers.gradleProperty("mcVersion")
+    filterPatches = false
     gitFilePatches = false
 
+    val fork = forks.register("folia") {
+        upstream.patchDir("paperServer") {
+            upstreamPath = "paper-server"
+            excludes = setOf("src/minecraft", "patches", "build.gradle.kts")
+            patchesDir = rootDirectory.dir("folia-server/paper-patches")
+            outputDir = rootDirectory.dir("paper-server")
+        }
+    }
+
+    activeFork = fork
+
+
     updatingMinecraft {
         // oldPaperCommit = "d4fe85375af18bfa88f44d7c1e6a61904ae550cc"
     }
@@ -86,7 +_,19 @@
     }
 }
 
-val log4jPlugins = sourceSets.create("log4jPlugins")
+sourceSets {
+    main {
+        java { srcDir("../paper-server/src/main/java"); srcDir("../paper-server/src/generated/java") }
+        resources { srcDir("../paper-server/src/main/resources") }
+    }
+    test {
+        java { srcDir("../paper-server/src/test/java") }
+        resources { srcDir("../paper-server/src/test/resources") }
+    }
+}
+val log4jPlugins = sourceSets.create("log4jPlugins") {
+    java { srcDir("../paper-server/src/log4jPlugins/java") }
+}
 configurations.named(log4jPlugins.compileClasspathConfigurationName) {
     extendsFrom(configurations.compileClasspath.get())
 }
@@ -109,7 +_,7 @@
 }
 
 dependencies {
-    implementation(project(":paper-api"))
+    implementation(project(":folia-api"))
     implementation("ca.spottedleaf:leafpile:1.0.0")
     implementation("org.jline:jline-terminal-ffm:3.27.1") // use ffm on java 22+
     implementation("org.jline:jline-terminal-jni:3.27.1") // fall back to jni on java 21
@@ -163,27 +_,29 @@
     manifest {
         val git = Git(rootProject.layout.projectDirectory.path)
         val mcVersion = rootProject.providers.gradleProperty("mcVersion").get()
-        val build = System.getenv("BUILD_NUMBER") ?: null
-        val buildTime = providers.environmentVariable("BUILD_STARTED_AT").map(Instant::parse).orElse(Instant.EPOCH).get()
+        val build = rootProject.providers.gradleProperty("tesseraBuildVersion").get() //System.getenv("BUILD_NUMBER") ?: "0040"
+        val buildTime = Instant.now()
         val gitHash = git.exec(providers, "rev-parse", "--short=7", "HEAD").get().trim()
         val implementationVersion = "$mcVersion-${build ?: "DEV"}-$gitHash"
         val date = git.exec(providers, "show", "-s", "--format=%ci", gitHash).get().trim()
         val gitBranch = git.exec(providers, "rev-parse", "--abbrev-ref", "HEAD").get().trim()
+        // tessera 26.2 start - change attributes for manifest
         attributes(
             "Main-Class" to "org.bukkit.craftbukkit.Main",
-            "Implementation-Title" to "Paper",
+            "Implementation-Title" to "Tessera",
             "Implementation-Version" to implementationVersion,
             "Implementation-Vendor" to date,
-            "Specification-Title" to "Paper",
+            "Specification-Title" to "Tessera",
             "Specification-Version" to project.version,
-            "Specification-Vendor" to "Paper Team",
-            "Brand-Id" to "papermc:paper",
-            "Brand-Name" to "Paper",
+            "Specification-Vendor" to "MosaikDev",
+            "Brand-Id" to "mosaikdev:tessera",
+            "Brand-Name" to "Tessera",
             "Build-Number" to (build ?: ""),
             "Build-Time" to buildTime.toString(),
             "Git-Branch" to gitBranch,
             "Git-Commit" to gitHash,
         )
+        // tessera 26.2 end - change attributes for manifest
         for (tld in setOf("net", "com", "org")) {
             attributes("$tld/bukkit", "Sealed" to true)
         }
@@ -316,7 +_,10 @@
 }
 
 fill {
-    project("paper")
+    // tessera 26.2 start - change project name
+//    project("folia")
+    project("tessera")
+    // tessera 26.2 end - change project name
     versionFamily(paperweight.minecraftVersion.map { it.split(".", "-").takeWhile { part -> part.toIntOrNull() != null }.take(2).joinToString(".") })
     version(paperweight.minecraftVersion)
 
