From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Hackii1432 <hunter122@hotmail.de>
Date: Thu, 30 Jul 2026 18:19:17 +0200
Subject: [PATCH] Adapt portal transitions to updated Paper API


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 0e02de971fdb02c04b09161a6eda3676ebf36135..a3d026243c776e67289b9e1cc2ba5bcb105420ec 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -4613,9 +4613,8 @@ public abstract class Entity
                                 new net.minecraft.world.level.portal.TeleportTransition(
                                     destination, Vec3.atBottomCenterOf(targetPos.below()), Vec3.ZERO, Direction.WEST.toYRot(), 0.0f,
                                     Relative.union(Relative.DELTA, Set.of(Relative.X_ROT)),
-                                    TeleportTransition.PLAY_PORTAL_SOUND.then(TeleportTransition.PLACE_PORTAL_TICKET),
-                                    org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.END_PORTAL
-                                )
+                                    TeleportTransition.PLAY_PORTAL_SOUND.then(TeleportTransition.PLACE_PORTAL_TICKET)
+                                ).withCause(org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.END_PORTAL)
                             );
                         }
                     );
@@ -4633,9 +4632,8 @@ public abstract class Entity
                                 new net.minecraft.world.level.portal.TeleportTransition(
                                     destination, Vec3.atBottomCenterOf(adjustedSpawn), Vec3.ZERO, 0.0f, 0.0f,
                                     Relative.union(Relative.DELTA, Relative.ROTATION),
-                                    TeleportTransition.PLAY_PORTAL_SOUND.then(TeleportTransition.PLACE_PORTAL_TICKET),
-                                    org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.END_PORTAL
-                                )
+                                    TeleportTransition.PLAY_PORTAL_SOUND.then(TeleportTransition.PLACE_PORTAL_TICKET)
+                                ).withCause(org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.END_PORTAL)
                             );
                         }
                     );
